### ISSUE TITLE
Class.LookupClass is slow

### ISSUE THREAD

**Issue by rolfbjarne** [xamarin/xamarin-macios#4936]
See profile here: https://github.com/xamarin/xamarin-macios/issues/4923#issuecomment-427175183, where Class.LookupClass is a significant portion of the timing.

The problem is that it ends up calling Assembly.GetName (), which recomputes the name every time, and is quite expensive to compute (as can be seen from the profile).

Ideas:

* Cache the IntPtr -> Type lookup in a dictionary.
* Make Assembly.GetName () cache the value (or cache it ourselves somehow).

**Comment by spouliot** [comment]
IIRC there was (before ref / corefx import) some caching in `Assembly.GetName ()` but it might have been wrong in some cases (ref emit). Still that might have caused a slow down.

It might make sense to cache it ourselves, even more if we're calling it inside other time sensitive code paths.

Still using a _lookup_ inside `Class.LookupClass` does make a lot of sense 😸

**Comment by rolfbjarne** [comment]
I've implemented 4 different optimizations (PRs #5009, #5013, #5016, #5017), which improved object creation by 69-86% (3-7x faster).

I also implemented a fifth optimization (create a factory class that can create object instances without using reflection), but it's much more invasive (and would bloat the assemblies a lot). The POC (with numbers) is [here](https://github.com/rolfbjarne/xamarin-macios/commit/c6c9e2a8ba2fe5ae8ca8b38027c84b902473fa7b). This optimization has good results (40% faster object creation), but due to its invasiveness I've discarded it for now.

**Comment by rolfbjarne** [comment]
I've implemented 4 different optimizations (PRs #5009, #5013, #5016, #5017), which improved object creation by 69-86% (3-7x faster).

I also implemented a fifth optimization (create a factory class that can create object instances without using reflection), but it's much more invasive (and would bloat the assemblies a lot). The POC (with numbers) is [here](https://github.com/rolfbjarne/xamarin-macios/commit/c6c9e2a8ba2fe5ae8ca8b38027c84b902473fa7b). This optimization has good results (40% faster object creation), but due to its invasiveness I've discarded it for now.
